### PR TITLE
Fix blueprint planner's validation of PlanningInput external networking

### PR DESCRIPTION
**Pull request by jgallagher**
The "Check the planning input" block of code needs to consider _all_ zones in the parent blueprint, including expunged zones. #6483 fixed the planner's ability to reuse external IPs from expunged zones, but accidentally made these checks incorrect, because it's certainly valid for the planning input to still have records for expunged zones. See https://github.com/oxidecomputer/omicron/pull/6581#discussion_r1765326105 for more context.

We also get to remove the somewhat-awkward `update_network_resources_from_blueprint()` test helper that was needed  to make some tests pass (because we didn't realize the checks being performed here were wrong).

### CODE REVIEW

File: nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs [hunk 1]
```diff
@@ -333,6 +276,139 @@ impl<'a> BuilderExternalNetworking<'a> {
     }
 }
 
+// Helper to validate that the system hasn't gone off the rails. There should
+// never be any external networking resources in the planning input (which is
+// derived from the contents of CRDB) that we don't know about from the parent
+// blueprint.
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs:282]: I might say something explaining that this _is_ possible, but means that our planning input has been invalidated, so the caller would be expected to take another lap.

File: nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs [hunk 2]
```diff
@@ -333,6 +276,139 @@ impl<'a> BuilderExternalNetworking<'a> {
...
+
+    // Unlike the construction of the external IP allocator and existing IPs
+    // constructed above in `BuilderExternalNetworking::new()`, we do not
+    // check for duplicates here: we could very well see reuse of IPs
+    // between expunged zones or between expunged -> running zones.
+    for (_, z) in parent_blueprint.all_omicron_zones(BlueprintZoneFilter::All) {
+        let zone_type = &z.zone_type;
+        match zone_type {
+            BlueprintZoneType::BoundaryNtp(ntp) => match ntp.nic.ip {
+                IpAddr::V4(ip) => {
```
Comment (davepacheco) [inline nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs:314]: Just a suggestion: could we make the sets contain `IpAddr`?  That'd consolidate it into fewer sets and eliminate the matching.

edit: I see that would make the code below a little more complicated.  I'll leave it to you to decide if it's any better on net.

**Review by davepacheco — APPROVED** [review]

File: nexus/reconfigurator/planning/src/blueprint_builder/builder.rs [hunk 3]
```diff
@@ -1592,6 +1593,28 @@ pub mod test {
         }
     }
 
+    #[track_caller]
+    pub fn assert_planning_makes_no_changes(
```
Comment (plotnick) [inline nexus/reconfigurator/planning/src/blueprint_builder/builder.rs:1597]: Awesome.

File: nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs (hunk 1 above)
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs]: Good call; expanded in [aca6f40](https://github.com/oxidecomputer/omicron/pull/6599/commits/aca6f40be8be6ca3c26e4ad802ef76e617446ae3)

File: nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs (hunk 2 above)
Comment (jgallagher) [reply nexus/reconfigurator/planning/src/blueprint_builder/external_networking.rs]: It's an improvement; the code below just needs some cheap `.into()`s ([9ae92dc](https://github.com/oxidecomputer/omicron/pull/6599/commits/9ae92dc9bbaeeadf149f23c26f3d640311c514b0)). Thanks!

**Review by plotnick — APPROVED** [review]
Looks great, thank you! Easy merge into #6581.

**Pull request merged**